### PR TITLE
Implement threaded execution for parameter update in staged optimizer

### PR DESCRIPTION
**Pull request by ttochi**
Based on `mpt+sso` branch (contributed by @gajagajago), modify some synchronization logic.

### CODE REVIEW

File: megatron/spiral/schedules.py
Comment (gajagajago) [inline megatron/spiral/schedules.py:1]: There is currently `offload_grad_curr` and `_offload_grad_ev_cpu` that have identical usage.  
Your `_offload_grad_ev_cpu` is passed to a newly spawned thread to synchronize before triggering SSO.
Original `offload_grad_curr` is for free stream to wait before calling init_context's `free_grad`. Original code doesn't work as intended when `free_grad` is simply setting .grad or .main_grad field to None, as that is a CPU job (as far as I know). 

I have two questions:
- Does your mechanism work?; passing torch cuda event made from a thread to another thread, and synchronizing at that thread? Is it checked to synchronize until the gradients are fully offloaded?
- Does this PR affect correctness of fp32 SSO?
- Can we just use one event for this, or is two events necessary?

File: megatron/training.py [hunk 1]
```diff
@@ -869,6 +869,7 @@ def train_step(forward_step_func, data_iterator,
                 model[_idx_bwd_stage_id].spiral_offload_grad(non_blocking=False) # blocking offload gradients
                 model[_idx_bwd_stage_id].spiral_free_grad()
         else:
+            torch.cuda.synchronize() # To ensure non_blocking offload finished
```
Comment (gajagajago) [inline megatron/training.py:872]: Is this necessary? We have optimizer joining related code in schedule.py, which waits for the SSO related operations to end. I think that is a superset of this torch.cuda.synchronize(), so I think this code is meaningless.

File: megatron/optimizer/optimizer.py [hunk 2]
```diff
@@ -476,7 +485,9 @@ def step(self, args, timers):
         # Step the optimizer.
         timers('optimizer-inner-step', log_level=1).start(
             barrier=args.barrier_with_L1_time)
-        self.optimizer.step()
+        self.optimizer.step(**inner_step_kwargs)
+        if hasattr(self.optimizer, "sync"):
+            self.optimizer.sync()
```
Comment (gajagajago) [inline megatron/optimizer/optimizer.py:490]: What does this code do? Is it adopted from somewhere else's code? ditto L842 of optimizer.py

**Review by gajagajago — COMMENTED** [review]
I left several reviews. Please address them. Thx for the PR.

File: megatron/training.py (hunk 1 above)
Comment (ttochi) [reply megatron/training.py:872]: This code line is for the case that `offload_grad_after_bwd_stage and not optimize_after_bwd_stage`
The detail explainment is in #25.

Comment (ttochi) [reply megatron/training.py:872]: Or maybe it would be better to move L886

File: megatron/optimizer/optimizer.py (hunk 2 above)
Comment (ttochi) [reply megatron/optimizer/optimizer.py:490]: This code waits for all parameter updates submitted to the thread in SpiralCPUAdam to finish.
I imported this part from `schedule.py` to `optimiser.step`.

File: megatron/spiral/schedules.py
Comment (ttochi) [reply megatron/spiral/schedules.py:1]: I am aware that the`offload_grad_curr` event and `free_grad` are not behaving as intended (#29) and it need to be fixed in another PR.

> Does your mechanism work?; passing torch cuda event made from a thread to another thread, and synchronizing at that thread? Is it checked to synchronize until the gradients are fully offloaded?

Yes I checked. This is profile result.
I add `offload-event-wait` nvtx in `_offload_grad_ev_cpu` event synchronization.
image

> Does this PR affect correctness of fp32 SSO?

It is not exactly the same as the GPU optimizer, but outputs similar loss.

> Can we just use one event for this, or is two events necessary?

~~It seems it can handle by one event! I will fix it :)~~
There's a conflict with something I'm working on after this PR. Can I fix it in another PR?

File: megatron/spiral/schedules.py [hunk 3]
```diff
@@ -906,13 +918,12 @@ def _cleanup():
                 == -1
             ):
                 raise RuntimeError("wait_event failed")
-
-            # join optimizer
-            if optimize_after_bwd_stage:
-                assert hasattr(getattr(optimizer[bwd_stage_id], "optimizer"), "sync"), "bwd_stage_optimizer.optimizer should be SpiralCPUAdam and have sync method"
-                optimizer[bwd_stage_id].optimizer.sync()
```
Comment (ttochi) [inline megatron/spiral/schedules.py:913]: https://github.com/mcrl/Megatron-LM-mcrl/pull/28#discussion_r1769601627

I moved this code block inside the SSO thread

File: megatron/spiral/schedules.py
Comment (gajagajago) [reply megatron/spiral/schedules.py:1]: Thx for clarifying.
If we want to use the original cuda event, we should remove the condition that only "free" stream can wait for it. 
Looking for the next PR to fix these issues!

File: megatron/training.py (hunk 1 above)
Comment (gajagajago) [reply megatron/training.py:872]: The code seems to be the case for only when `offload_grad_after_bwd_stage`. 

```
def train_step():

   ... skip ...

    # Offload gradients.
    if args.spiral:
        if not offload_grad_after_bwd_stage:
            for bwd_stage_id in range(mpu.get_spiral_backward_virtual_size()):
                _idx_bwd_stage_id = -bwd_stage_id - 1 if args.spiral_remap else bwd_stage_id
                model[_idx_bwd_stage_id].spiral_offload_grad(non_blocking=False) # blocking offload gradients
                model[_idx_bwd_stage_id].spiral_free_grad()
        else:
            torch.cuda.synchronize() # To ensure non_blocking offload finished
            # Assert gradients are already offloaded and freed on GPU
            for bwd_stage_id in range(mpu.get_spiral_backward_virtual_size()):
                _idx_bwd_stage_id = -bwd_stage_id - 1 if args.spiral_remap else bwd_stage_id
                model[_idx_bwd_stage_id].spiral_assert_free_grad()

```

**Comment by gajagajago** [comment]
With other comments resolved, I left one extra comment regarding training.py L872. Please check!

File: megatron/training.py (hunk 1 above)
Comment (ttochi) [reply megatron/training.py:872]:
```
    offload_grad_after_bwd_stage = args.spiral_overlap_offload_grad
    optimize_after_bwd_stage = args.spiral_overlap_offload_grad and args.spiral_stage_optimizer
```
it can be entered when `spiral_overlap_offload_grad and not spiral_stage_optimizer`

Comment (gajagajago) [reply megatron/training.py:872]: Okay, but I think synchronizing the entire device for ensuring grad offload function is bad pattern. I will refactor this very soon.
